L2 genesis - refactor to use solidity script

protolambda opened the pull request:
**Description**

Rebased version of https://github.com/ethereum-optimism/optimism/pull/9792

Taking ownership over this PR to add ProxyAdmin, Preinstalls, and integrate into op-e2e / devnet testing.

TODO:
- [x] Work from Mark to revive and improve changes from Wyatt to add missing predeploys to the script.
- [x] Insert `ProxyAdmin` into L2 predeploys
- [x] Support Preinstalls in solidity script
  - [x] `Permit2`
  - [x] `Create2Deployer`
  - [x] `DeterministicDeploymentProxy`
  - [x] `EntryPoint (4337)`
  - [x] `Safe_v130`
  - [x] `SenderCreator`
  - [x] `SafeSingletonFactory`
  - [x] `MultiSend_v130`
  - [x] `MultiCall3`
  - [x] `SafeL2_v130`
  - [x] `MultiSendCallOnly_v130`
- [x] Apply ecotone
- [x] Update `Setup` solidity contract to use this, rather than Go FFI
- [ ] ~~Re-use superchain-registry code path to build full `genesis.json` and `rollup.json``~~ Minimized op-chain-ops a lot already, reusing registry requires more diff, probably better for a different PR.
- [x] Integrate into op-e2e
- [x] Integrate into devnet (running in CI, verifying results still)

**Metadata**

Fix https://github.com/ethereum-optimism/protocol-quest/issues/151

Spacesai1or committed:
L2 genesis solidity updates by Wyatt and Mark

Commits:
- Tmp change for Graphite
- Add Missing Predeploys to L2 Genesis Script
- contracts-bedrock: refactor L2 genesis generation
- wip
- temp
- l2 genesis generation wip
- updates

Co-authored-by: Mark Tyneway <mark.tyneway@gmail.com>
Co-authored-by: Wyatt Barnes <me@wyatt.email>

protolambda committed:
L2 genesis refactor fixes

Squashed commits:

todo

add ProxyAdmin predeploy, clean up test assertions

order predeploy setters

split Predeploys and Preinstalls

L2 genesis script: bytes32(0) style suggestion

move predeploy utils to predeploy library

preinstalls bytecode

fix typos

permit2 bytecode immutable patching

activate ecotone

cleanup tests, fixes, ecotone

work in progress GenesisL2 addition to Setup()

fixes

devnet allocs

fix go lint

more fixes

fix predeploys proxy impl checking test

fix solady

continue Go integration

l2 genesis Go integration stuff

fix lint

fix go lint

lint fixes

fixes for some PR review comments

Predeploys test clean up

go test superseded by solidity testing

fix lint

cleanup and review fixes

minor fixes

test fixes

op-e2e l2 allocs filepath fix

more devnet test funds

improve logging

fix devnet allocs-l2 path naming and output file moving to .devnet

devnet allocs CI fixes

circle ci workspace allocs-l2 fixes

op-e2e: fix alloc npe

fix

enforce genesis allocs copy

op-e2e fix

go test fixes, 4788 nonce edge case, dev accounts fix, misc fixes

fix test, fix flake

fix tests

Proxy artifact workaround

update gas snapshot

undo workaround, apply config change to fix

undo failed workarounds

tynes commented on packages/contracts-bedrock/scripts/L2Genesis.s.sol:
```diff
@@ -123,66 +124,178 @@ contract L2Genesis is Script, Artifacts {
+        // Note the ProxyAdmin implementation itself is behind a proxy that owns itself.
+        _setImplementationCode(Predeploys.PROXY_ADMIN);
+
+        bytes32 _ownerSlot = hex"0000000000000000000000000000000000000000000000000000000000000000";
```
You can represent this as `bytes32(0)` more cleanly

protolambda replied:
I thought about it, but this is consistent with all the other slot declarations. I'll change it to `bytes32(0)` if you like that better though.

tynes replied:
I think it would be much cleaner to do `bytes32(0)`, less characters, easier to update + verify correctness, the old stuff was just there and should be updated to new syntax if necessary as well

tynes commented on op-node/cmd/genesis/cmd.go:
```diff
@@ -165,6 +170,16 @@ var Subcommands = cli.Commands{
 			}
 
+			var dump *state.Dump
+			if l2Allocs := ctx.String("l2-allocs"); l2Allocs != "" {
```
This should be required?

tynes commented on packages/contracts-bedrock/src/libraries/Predeploys.sol:
```diff
@@ -69,45 +69,88 @@ library Predeploys {
+        if (_addr == EAS) { return "EAS"; }
+        if (_addr == GOVERNANCE_TOKEN) { return "GovernanceToken"; }
+        if (_addr == LEGACY_ERC20_ETH) { return "LegacyERC20ETH"; }
+        revert("unnamed predeploy");
```
needs a prefix of the contract name to fix this like `Predeploys: unnamed predeploy`

tynes commented on packages/contracts-bedrock/scripts/L2Genesis.s.sol:
```diff
@@ -74,46 +74,61 @@ contract L2Genesis is Script, Artifacts {
 
     /// @dev Reads the deploy config, sets `outfile` which is where the `vm.dumpState` will be saved to, and
     ///      loads in the addresses for the L1 contract deployments.
     function setUp() public override {
```
you can technically delete this `setUp` as it does nothing in addition to the `super.setUp`

protolambda commented:
rebased on develop, fixed an ugly few solidity test integration issues, devnet should work (will watch for CI) and going to work on op-e2e now

tynes commented on packages/contracts-bedrock/test/L2Genesis.t.sol:
```diff
@@ -0,0 +1,263 @@
+    }
+
+    // this is slower..
+    function getBalance(string memory account) internal returns (uint256) {
```
We should delete the extra unused functions in this file

protolambda replied:
Maybe we should add some addiitonal checks using these functions instead? What ideas did you have for testing in the original draft?

tynes replied:
The problem was that they were prohibitively slow

tynes replied:
It will take forever to run if we do something like "iterate over each account and get its balance" so we should say "jq get all accounts that have non zero balance" and "jq get all accounts that have 1 wei" and assert that the number is the same, this will say there are definitely only that number of accounts with 1 wei and it should be the expected number which iirc is 256

tynes commented on packages/contracts-bedrock/test/L2Genesis.t.sol:
```diff
@@ -0,0 +1,263 @@
+        return abi.decode(vm.ffi(commands), (uint256));
+    }
+
+    // can this become a modifier?
```
dead comment

tynes commented on op-chain-ops/genesis/layer_two_test.go:
```diff
@@ -131,73 +130,72 @@ func testBuildL2Genesis(t *testing.T, config *genesis.DeployConfig) *core.Genesi
-	require.Equal(t, big.NewInt(1), blobBaseFee)
+	require.Equal(t, big.NewInt(0).String(), blobBaseFee.String())
 }
 
```
We can delete this test in the future, the solidity tests should cover this

tynes commented on op-chain-ops/genesis/layer_two_test.go:
```diff
@@ -34,7 +33,7 @@ func init() {
 var testKey, _ = crypto.HexToECDSA("b71c71a67e1177ad4e901695e1b4b9ee17ae16c6668d313eac2f96dbcda3f291")
 
 // Tests the BuildL2MainnetGenesis factory with the provided config.
```
Its good to leave this test for now but we should def move away from it in the future and rely completely on the solidity test coverage which i do believe almost already covers this besides the "use gov token" flag

tynes commented on packages/contracts-bedrock/scripts/Config.sol:
```diff
@@ -94,7 +95,7 @@ library Config {
             return "optimism-goerli";
```
In the future we should move away from using this function but fine to keep for now as it would be a breaking change, its just not easy to maintain and also results in implicit behavior that is non standard based on chain id as not all chain ids are considered "known"

tynes commented on packages/contracts-bedrock/scripts/L2Genesis.s.sol:
```diff
@@ -74,46 +73,55 @@ contract L2Genesis is Script, Artifacts {
+    function runWithOptions(OutputMode _mode, L1Dependencies memory _l1Dependencies) public {
+        dealEthToPrecompiles();
+        setPredeployProxies();
+        setPredeployImplementations(_l1Dependencies);
```
This creates a really good way for us to standardize on the initial release of bedrock inside of `setPredeployImplementations`, we can copy/paste the initial release bytecode to be used there. This would mean we want a "development" output mode which uses the contracts based on the source code when the command runs. No need to do this right now but curious what you think of this approach?

tynes replied:
Ah this dev mode is local delta

tynes commented on packages/contracts-bedrock/scripts/L2Genesis.s.sol:
```diff
@@ -123,66 +131,170 @@ contract L2Genesis is Script, Artifacts {
+        vm.etch(impl, address(vault).code);
+
+        /// Reset so its not included state dump
+        vm.etch(address(vault), "");
```
it would probably be good to turn the reset into an internal function and reuse its impl everywhere. You can see how forge-std does it [here](https://github.com/foundry-rs/forge-std/blob/5dd1c68131ddd3c89ef169666eb262b92e90507c/src/StdCheats.sol#L547)

we don't need to go as far as they do, for us setting the balance to 0 technically isn't required as we don't have balance in this context

protolambda replied:
This is not a full reset, but rather the reset of just the specific contract that was temporarily deployed for genesis purposes. Open to suggestions for how the reset should be, but don't think there's a lot to reduce.

tynes commented on packages/contracts-bedrock/scripts/L2Genesis.s.sol:
```diff
@@ -318,112 +389,103 @@ contract L2Genesis is Script, Artifacts {
-        address impl = _predeployToCodeNamespace(_addr);
-        console.log("Setting %s implementation at: %s", _name, impl);
-        vm.etch(impl, vm.getDeployedCode(string.concat(_name, ".sol:", _name)));
+    /// @dev Sets all the preinstalls.
```
small semgrep fixes

tynes commented on packages/contracts-bedrock/scripts/L2Genesis.s.sol:
```diff
@@ -318,112 +389,103 @@ contract L2Genesis is Script, Artifacts {
+        bool hasBeaconBlockRoots = false;
+        address addr = Preinstalls.BeaconBlockRoots;
+        assembly {
+            hasBeaconBlockRoots := extcodesize(addr)
```
You can do `address.code.length`, no need for assembly

tynes commented on packages/contracts-bedrock/scripts/L2Genesis.s.sol:
```diff
@@ -318,112 +389,103 @@ contract L2Genesis is Script, Artifacts {
+        assembly {
+            hasBeaconBlockRoots := extcodesize(addr)
+        }
+        require(hasBeaconBlockRoots, "must have beacon-block-roots contract");
```
semgrep wants the contract name prefixed like so `L2Genesis: ...`

tynes commented on packages/contracts-bedrock/scripts/L2Genesis.s.sol:
```diff
@@ -318,112 +389,103 @@ contract L2Genesis is Script, Artifacts {
+    }
 
-        _verifyProxyImplementationAddress(_addr, impl);
+    function activateEcotone() public {
```
This technically doesn't fully activate ecotone, we would be able to make good use of https://github.com/foundry-rs/foundry/pull/4931 to accomplish this easily. There are a few paths forward here:

- get the cheatcode into forge and just hardcode the rlp encoded tx. This would also require building foundry with optimism feature flag since its a deposit tx
- defining a solidity version of the "message" that is represented by the upgrade deposit tx

I lean towards just defining the messages given we would need to use a special foundry with optimism build flag and its unclear if that would even work e2e. Not ideal that its not 1:1 with the go, we would prob want a differential test long term

protolambda replied:
The upgrade txs don't apply, as it deploys the same contracts that are already in place by default. The 4788 contract is installed as preinstall, independent of hardfork. And then only the `setL1BlockValuesEcotone` call remains to do here.

tynes commented on packages/contracts-bedrock/test/Preinstalls.t.sol:
```diff
@@ -0,0 +1,110 @@
+        assertEq(domainSeparator, expectedDomainSeparator, "Domain separator mismatch");
+        assertEq(chainId, uint256(31337));
+        assertEq(domainSeparator, bytes32(0x4d553c58ae79a6c4ba64f0e690a5d1cd2deff8c6b91cf38300e0f2b76f9ee346));
+        // Warning the Permit2 domain separator as cached in the DeployPermit2.sol bytecode is incorrect.
```
Can't we update it? hmm

protolambda replied:
Bug in the upstream library test tooling that has existed forever. Don't think we need to change it. The mainnet L1 bytecode serves as a template, that passes tests when using chainID=1 itself, well enough.

tynes commented on packages/contracts-bedrock/test/Preinstalls.t.sol:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    function test_preinstall_beaconBlockRoots_succeeds() external {
+        // TODO: this test fails with legacy genesis, due to missing 4788 contract there (no ecotone support).
```
We could set the deploy type to latest to fix this?

tynes commented on packages/contracts-bedrock/scripts/L2Genesis.s.sol:
```diff
@@ -74,46 +73,55 @@ contract L2Genesis is Script, Artifacts {
+        }
+        if (_mode == OutputMode.OUTPUT_ALL) {
+            writeGenesisAllocs(Config.stateDumpPath("-delta"));
+        }
```
How would we express just ecotone? Should we make mode more of a chainspec thing where its the hardfork name?

protolambda replied:
ecotone is the default, latest fork, without any suffix.

tynes commented on packages/contracts-bedrock/scripts/L2Genesis.s.sol:
```diff
@@ -74,46 +73,55 @@ contract L2Genesis is Script, Artifacts {
+        return L1Dependencies({
+            deployedL1CrossDomainMessengerProxy: mustGetAddress("L1CrossDomainMessengerProxy"),
+            deployedL1StandardBridgeProxy: mustGetAddress("L1StandardBridgeProxy"),
+            deployedL1ERC721BridgeProxy: mustGetAddress("L1ERC721BridgeProxy")
```
smol nit: i don't think we need the `deployed` prefix, could just add a comment on the function name

tynes commented on packages/contracts-bedrock/src/libraries/Predeploys.sol:
```diff
@@ -69,45 +68,73 @@ library Predeploys {
 
-    /// @notice Address of the EntryPoint predeploy.
-    address internal constant EntryPoint = 0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789;
+    /// @notice Returns the name of the predeploy at the given address.
```
I like this function and it can be used in multiple places

tynes commented:
Perhaps we want to improve the log `retrieving L1 deployments from artifacts` so it includes the path that its reading from? Looks like CI is failing on this right now

tynes commented on packages/contracts-bedrock/src/libraries/Predeploys.sol:
```diff
@@ -69,45 +68,75 @@ library Predeploys {
 
-    /// @notice Address of the EntryPoint predeploy.
-    address internal constant EntryPoint = 0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789;
+    /// @notice Returns the name of the predeploy at the given address.
+    function getName(address _addr) internal pure returns (string memory out_) {
+        require(isPredeployNamespace(_addr), "Predeploys: address must be a predeploy");
+        if (_addr == LEGACY_MESSAGE_PASSER) return "LegacyMessagePasser";
+        if (_addr == L1_MESSAGE_SENDER) return "L1MessageSender";
+        if (_addr == DEPLOYER_WHITELIST) return "DeployerWhitelist";
+        if (_addr == WETH9) return "WETH9";
+        if (_addr == L2_CROSS_DOMAIN_MESSENGER) return "L2CrossDomainMessenger";
+        if (_addr == GAS_PRICE_ORACLE) return "GasPriceOracle";
+        if (_addr == L2_STANDARD_BRIDGE) return "L2StandardBridge";
+        if (_addr == SEQUENCER_FEE_WALLET) return "SequencerFeeVault";
+        if (_addr == OPTIMISM_MINTABLE_ERC20_FACTORY) return "OptimismMintableERC20Factory";
+        if (_addr == L1_BLOCK_NUMBER) return "L1BlockNumber";
+        if (_addr == L2_ERC721_BRIDGE) return "L2ERC721Bridge";
+        if (_addr == L1_BLOCK_ATTRIBUTES) return "L1Block";
+        if (_addr == L2_TO_L1_MESSAGE_PASSER) return "L2ToL1MessagePasser";
+        if (_addr == OPTIMISM_MINTABLE_ERC721_FACTORY) return "OptimismMintableERC721Factory";
+        if (_addr == PROXY_ADMIN) return "ProxyAdmin";
+        if (_addr == BASE_FEE_VAULT) return "BaseFeeVault";
+        if (_addr == L1_FEE_VAULT) return "L1FeeVault";
+        if (_addr == SCHEMA_REGISTRY) return "SchemaRegistry";
+        if (_addr == EAS) return "EAS";
+        if (_addr == GOVERNANCE_TOKEN) return "GovernanceToken";
+        if (_addr == LEGACY_ERC20_ETH) return "LegacyERC20ETH";
+        revert("Predeploys: unnamed predeploy");
+    }
```
Is that valid solidity?

protolambda commented:
@tynes there are already sufficient logs / forge debug data to see the exact issue:
```
    ├─ [0] VM::getDeployedCode("Proxy.sol:Proxy") [staticcall]
    │   └─ ← [Revert] No matching artifact found
    └─ ← [Revert] No matching artifact found
```
It fails to retrieve the `Proxy` artifact.

protolambda commented on packages/contracts-bedrock/scripts/L2Genesis.s.sol:
```diff
@@ -71,49 +72,65 @@ contract L2Genesis is Script, Artifacts {
+    function runWithOptions(OutputMode _mode, L1Dependencies memory _l1Dependencies) public {
+        dealEthToPrecompiles();
+        setPredeployProxies();
+        setPredeployImplementations(_l1Dependencies);
```
bad suggestion

protolambda commented:
I think the `Proxy.sol` not found error, even though it shows up in the artifacts directory, is related to: https://github.com/foundry-rs/foundry/issues/7569

trianglesphere commented on op-chain-ops/genesis/layer_two.go:
```diff
@@ -1,140 +1,54 @@
-			return nil, fmt.Errorf("code not set for %s", name)
-		}
+	genspec.Alloc = dump.Accounts
+	// sanity check the permit2 immutable, to verify we using the allocs for the right chain.
```
How confident are we that the chainID will always be at this bytecode location?

protolambda replied:
100%, the bytecode of permit2 is hardcoded, not compiled from source. And permit2 itself is not changing anymore either. Solidity unit tests ensure that the behavior is correct with different chain IDs at this offset.

0xfuturistic reviewed (approved):
lgtm after resolving conflict

protolambda commented:
Rebased + fixed devnet python merge conflict + added gas snapshot + added Proxy import into L2Genesis script to try and fix missing artifact flake / forge issue.

protolambda commented:
Blocked by https://github.com/foundry-rs/foundry/issues/7732

protolambda commented:
Rebased and squashed my commits, `forge script` fix should now work in CI. Really looking forward to get this refactor merged. Will be ooo this week otherwise.

tynes commented:
Accidentally closed this with my commit message

tynes committed:
contracts-bedrock: delete dead L2 genesis testing code

tynes committed:
state-diff: update

tynes committed:
ctb: L2 genesis delete dead comment

tynes committed:
contracts-bedrock: small cleanups

tynes committed:
ctb: cleanup L2 genesis comments
